File: server/lib/ollama/ollama_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// Minimal Ollama HTTP client. Targets the `/api/chat` endpoint, which
/// supports structured tool calls on recent server versions.
class OllamaClient {
  OllamaClient({required Uri baseUrl, required this.model, http.Client? client})
      : _baseUrl = baseUrl,
        _client = client ?? http.Client();

  final Uri _baseUrl;
  final String model;
  final http.Client _client;

  void close() => _client.close();

  /// Non-streaming chat completion. [tools] is the JSON-schema tool list
  /// in Ollama's format; pass `null` to disable tool calling.
  Future<OllamaChatResponse> chat({
    required List<OllamaMessage> messages,
    List<Map<String, Object?>>? tools,
    Map<String, Object?>? format,
    double temperature = 0.2,
  }) async {
    final url = _baseUrl.resolve('/api/chat');
    final body = <String, Object?>{
      'model': model,
      'stream': false,
      'messages': messages.map((m) => m.toJson()).toList(),
      'options': {'temperature': temperature},
      if (tools != null) 'tools': tools,
      if (format != null) 'format': format,
    };
    final resp = await _client.post(
      url,
      headers: {'content-type': 'application/json'},
      body: jsonEncode(body),
    );
    if (resp.statusCode >= 400) {
      throw OllamaException(
        'Ollama ${resp.statusCode}: ${resp.body}',
        statusCode: resp.statusCode,
      );
    }
    final json = jsonDecode(resp.body) as Map<String, Object?>;
    return OllamaChatResponse.fromJson(json);
  }
}

class OllamaMessage {
  const OllamaMessage({
    required this.role,
    required this.content,
    this.toolName,
    this.toolCalls,
  });

  final String role; // system | user | assistant | tool
  final String content;
  final String? toolName;
  final List<OllamaToolCall>? toolCalls;

  Map<String, Object?> toJson() => {
        'role': role,
        'content': content,
        if (toolName != null) 'name': toolName,
        if (toolCalls != null)
          'tool_calls': toolCalls!.map((c) => c.toJson()).toList(),
      };
}

class OllamaToolCall {
  const OllamaToolCall({required this.name, required this.arguments});
  final String name;
  final Map<String, Object?> arguments;

  factory OllamaToolCall.fromJson(Map<String, Object?> json) {
    final function = (json['function'] as Map?)?.cast<String, Object?>() ?? {};
    final rawArgs = function['arguments'];
    final Map<String, Object?> args = switch (rawArgs) {
      Map() => rawArgs.cast<String, Object?>(),
      String() => jsonDecode(rawArgs) as Map<String, Object?>,
      _ => <String, Object?>{},
    };
    return OllamaToolCall(
      name: function['name'] as String? ?? '',
      arguments: args,
    );
  }

  Map<String, Object?> toJson() => {
        'function': {'name': name, 'arguments': arguments},
      };
}

class OllamaChatResponse {
  OllamaChatResponse({
    required this.content,
    required this.toolCalls,
    required this.done,
  });

  final String content;
  final List<OllamaToolCall> toolCalls;
  final bool done;

  factory OllamaChatResponse.fromJson(Map<String, Object?> json) {
    final message = (json['message'] as Map?)?.cast<String, Object?>() ?? {};
    final rawCalls = (message['tool_calls'] as List?) ?? const [];
    return OllamaChatResponse(
      content: (message['content'] as String?) ?? '',
      toolCalls: rawCalls
          .cast<Map>()
          .map((m) => OllamaToolCall.fromJson(m.cast<String, Object?>()))
          .toList(),
      done: (json['done'] as bool?) ?? true,
    );
  }
}

class OllamaException implements Exception {
  OllamaException(this.message, {this.statusCode});
  final String message;
  final int? statusCode;
  @override
  String toString() => 'OllamaException: $message';
}

File: server/app/agent/ollama_client.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

import httpx


@dataclass
class OllamaMessage:
    role: str  # system | user | assistant | tool
    content: str
    tool_name: str | None = None
    tool_calls: list[OllamaToolCall] | None = None

    def to_json(self) -> dict[str, Any]:
        out: dict[str, Any] = {"role": self.role, "content": self.content}
        if self.tool_name is not None:
            out["name"] = self.tool_name
        if self.tool_calls is not None:
            out["tool_calls"] = [c.to_json() for c in self.tool_calls]
        return out


@dataclass
class OllamaToolCall:
    name: str
    arguments: dict[str, Any] = field(default_factory=dict)

    def to_json(self) -> dict[str, Any]:
        return {"function": {"name": self.name, "arguments": self.arguments}}

    @classmethod
    def from_json(cls, raw: dict[str, Any]) -> OllamaToolCall:
        fn = raw.get("function") or {}
        args_raw = fn.get("arguments")
        if isinstance(args_raw, dict):
            args = args_raw
        elif isinstance(args_raw, str) and args_raw.strip():
            try:
                args = json.loads(args_raw)
            except json.JSONDecodeError:
                args = {}
        else:
            args = {}
        return cls(name=fn.get("name", ""), arguments=args)


@dataclass
class OllamaChatResponse:
    content: str
    tool_calls: list[OllamaToolCall]
    done: bool

    @classmethod
    def from_json(cls, raw: dict[str, Any]) -> OllamaChatResponse:
        msg = raw.get("message") or {}
        tool_calls_raw = msg.get("tool_calls") or []
        return cls(
            content=msg.get("content", ""),
            tool_calls=[OllamaToolCall.from_json(c) for c in tool_calls_raw],
            done=raw.get("done", True),
        )


class OllamaError(RuntimeError):
    def __init__(self, message: str, status_code: int | None = None) -> None:
        super().__init__(message)
        self.status_code = status_code


class OllamaClient:
    """Minimal Ollama HTTP client targeting ``/api/chat``."""

    def __init__(self, base_url: str, model: str, timeout: float = 300.0) -> None:
        self.base_url = base_url.rstrip("/")
        self.model = model
        self._client = httpx.AsyncClient(timeout=timeout)

    async def close(self) -> None:
        await self._client.aclose()

    async def chat(
        self,
        *,
        messages: list[OllamaMessage],
        tools: list[dict[str, Any]] | None = None,
        response_format: dict[str, Any] | None = None,
        temperature: float = 0.2,
    ) -> OllamaChatResponse:
        body: dict[str, Any] = {
            "model": self.model,
            "stream": False,
            "messages": [m.to_json() for m in messages],
            "options": {"temperature": temperature},
        }
        if tools is not None:
            body["tools"] = tools
        if response_format is not None:
            body["format"] = response_format

        resp = await self._client.post(f"{self.base_url}/api/chat", json=body)
        if resp.status_code >= 400:
            raise OllamaError(
                f"Ollama {resp.status_code}: {resp.text}",
                status_code=resp.status_code,
            )
        return OllamaChatResponse.from_json(resp.json())
